packages/cloudfront-sign/src/sign.ts

```diff
@@ -0,0 +1,203 @@
+    const cidr = parseCIDR(args.ipAddress);
+    policy.Statement[0].Condition["IpAddress"] = {
+      "AWS:SourceIp": cidr,
+    };
+  }
+  return policy;
+}
+
+function signData(data: string, privateKeyBuffer: Buffer): string {
+  const sign = createSign("RSA-SHA1");
+  sign.update(data);
+  return sign.sign(privateKeyBuffer, "base64");
+}
+
+function parseDate(date?: string): number | undefined {
+  if (!date) {
+    return undefined;
+  }
+  const parsedDate = new Date(date);
+  return parsedDate instanceof Date ? epochTime(parsedDate) : undefined;
+}
+
+function parseDateWindow(expiration: string, start?: string): PolicyDates {
+  const dateLessThan = parseDate(expiration);
+  if (!dateLessThan) {
+    throw new Error("dateLessThan is invalid. Ensure a valid date string is provided.");
+  }
+  return {
+    dateLessThan,
+    dateGreaterThan: parseDate(start),
+  };
+}
+
+function signPolicy(policy: string, privateKey: string): string {
+  const privateKeyBuffer = readFileSync(privateKey);
+  return normalizeBase64(signData(policy, privateKeyBuffer));
+}
+
+export function signUrl(args: CloudfrontSignInput): string {
+  const { dateLessThan, dateGreaterThan } = parseDateWindow(args.dateLessThan, args.dateGreaterThan);
+  const url = parseUrl(args.url);
+  const policy = JSON.stringify(
+    buildPolicy({
+      dateLessThan,
+      dateGreaterThan,
+      ipAddress: args.ipAddress,
+      resource: args.url,
+    })
+  );
+  const signature = signPolicy(policy, args.privateKey);
+  const cloudfrontQueryParams = [];
+  for (const key in url.query) {
+    cloudfrontQueryParams.push(`${key}=${url.query[key]}`);
+  }
+  const usingACustomPolicy = Boolean(dateGreaterThan) || Boolean(args.ipAddress);
+  if (usingACustomPolicy) {
+    const base64EncodedPolicy = encodeToBase64(policy);
+    cloudfrontQueryParams.push(`Policy=${base64EncodedPolicy}`);
+  } else {
+    cloudfrontQueryParams.push(`Expires=${dateLessThan}`);
+  }
+  cloudfrontQueryParams.push(`Key-Pair-Id=${args.keyPairId}`);
+  cloudfrontQueryParams.push(`Signature=${signature}`);
+  return `${args.url.split("?")[0]}?${cloudfrontQueryParams.join("&")}`;
+}
+
+export function signCookies(args: CloudfrontSignInput): CloudfrontSignedCookiesOutput {
+  const { dateLessThan, dateGreaterThan } = parseDateWindow(args.dateLessThan, args.dateGreaterThan);
+  const usingACustomPolicy = Boolean(dateGreaterThan) || Boolean(args.ipAddress);
+  const policy = JSON.stringify(
+    buildPolicy({
+      dateLessThan,
+      dateGreaterThan,
+      ipAddress: args.ipAddress,
+      resource: args.url,
+    })
+  );
+  const signature = signPolicy(policy, args.privateKey);
+  const base64EncodedPolicy = encodeToBase64(policy);
+  return {
+    "CloudFront-Key-Pair-Id": args.keyPairId,
+    "CloudFront-Signature": signature,
+    "CloudFront-Expires": !usingACustomPolicy ? dateLessThan : undefined,
+    "CloudFront-Policy": usingACustomPolicy ? base64EncodedPolicy : undefined,
+  };
+}
```

Comment:
overall: there are two things we may want to incorporate from the code snippet by @ffxsam here: https://github.com/aws/aws-sdk-js-v3/issues/1822#issuecomment-882900924

- One is the bifurcated exclusive typing for policy and expires used in the interfaces
- RTMP support in the implementation

Reply:
Okay. I understand the need to introduce parts of this snippet. RTMP support is straight forward. I have a few comments about the first request.

> One is the bifurcated exclusive typing for policy and expires used in the interfaces

This is a good idea. Being explicit with a union type for the parameter makes it clear what arguments the user can and will need to provide.

The current implementation uses the name `dateLessThan`, in the snippet it uses `expires`, and it was influenced by the AWS CLI [Cloudfront.sign](https://awscli.amazonaws.com/v2/documentation/api/latest/reference/cloudfront/sign.html) function. The decision behind this name `dateLessThan` was because you could also provide a start date to a custom policy. In the current implementation this start date is named `dateGreaterThan`. What are your thoughts to keep it named as it is?

Additionally, I foresee at least three possible use cases to these signing functions:
1. "I am a user that wants to sign a URL with an expiration date"
1. "I am a user that wants to sign a URL with a custom policy I defined"
1. "I am a user that wants to sign a URL with an expiration date, and possibly other conditions such as only allowing access by a specific IP address or only allowing access after a date"

The snippet covers the first two cases. For case three I suspect some users don't have the knowledge or desire to create a custom policy and would rather provide the signing functions the url and expiration date, and optional ip address and start date conditions. Then this package would create, apply, and sign the custom policy for them. The current implementation is already doing this.

If I leverage the union type from the snippet (`GetSignedUrlOptions`) it could consist of three types:
1. one that requires a url and expiration date (`GetSignedUrlWithExpires`)
1. one that requires a url and a policy (`GetSignedUrlWithPolicy`)
1. one that requires a url and expiration date, and optional IP address and start date.

What are your thoughts around this new type? How do you envision this being defined?

Reply:
Can do it with either runtime check or TypeScript:

```typescript
export type CloudfrontSignInput = CloudfrontSignInputWithParameters | CloudfrontSignInputWithPolicy

export type CloudfrontSignInputWithParameters = {
  url: string;
  keyPairId: string;
  privateKey: string | Buffer;
  dateLessThan: string;
  ipAddress?: string;
  dateGreaterThan?: string;
  // policy?: never
}

export type CloudfrontSignInputWithPolicy = {
  url: string;
  keyPairId: string;
  privateKey: string | Buffer;
  policy: string;
  // dateLessThan?: never, ipAddress?: never, dateGreaterThan?: never
}
```
Options:

- set them to `?: never` in the other type, like in https://github.com/aws/aws-sdk-js-v3/issues/1822#issuecomment-882900924
- throw an illegal argument error if you detect both policy and other params
- do nothing, take policy as the priority field (not recommended)